File: AOJ/0260/0260.cpp
#include<iostream>
#include<algorithm>
using namespace std;
static const int MAX = 65000;
typedef long long llong;

void merge(long long *J, int l, int m, int r){
  long long T[MAX];
  int i, j, k;
  for ( i = l; i < m; i++ ) T[i] = J[i];
  for ( i = m; i < r; i++ ) T[(r - 1) - (i - m)] = J[i];
  i = l;
  j = r - 1;
  for ( k = l; k < r; k++ ){
    if ( T[i] <= T[j] ) J[k] = T[i++];
    else J[k] = T[j--];
  }
}

void mergeSort(long long *J, int l, int r){
  if ( l + 1 >= r ) return;
  int m = (l + r) / 2;
  mergeSort(J, l, m);
  mergeSort(J, m, r);
  merge(J, l, m, r);
}

int main(){
  llong n, totalp, totalj;
  llong P[MAX], J[MAX];

  while(1){
    cin >> n;
    if ( n == 0 ) break;
    totalp = 0;
    for ( int i = 0; i < n; i++ ) {
      cin >> P[i];
      totalp += P[i];
    }
    for ( int i = 0; i < n - 1; i++ ) cin >> J[i];

    mergeSort(J, 0, n - 1);
    J[n - 1] = 0;
    
    llong maxv = 0;
    totalj = 0;
    for ( int k = 0; k <= n - 1; k++ ){
      totalj += J[n - 1 - k];
      maxv = max((n - k) * (totalj + totalp), maxv);
    }
    cout << maxv << endl;
  }
}

File: AOJ/0260/0260.py
MAX = 65000
T = [0] * MAX
def merge(J, l, m, r):
    for i in range(l, m): T[i] = J[i]
    for i in range(m, r): T[(r -1) - (i - m)] = J[i]
    i = l
    j = r - 1
    for k in range(l, r):
        if  T[i] <= T[j]: J[k] = T[i]; i += 1
        else: J[k] = T[j]; j -= 1

def mergeSort(J, l, r):
    if l + 1 >= r: return
    m = (l + r) // 2
    mergeSort(J, l, m)
    mergeSort(J, m, r)
    merge(J, l, m, r)

while True:
    n = int(input())
    if n == 0: break

    totalp = 0
    P = list(map(int, input().split()))
    for p in P:
        totalp += p
    J = list(map(int, input().split()))

    maxv = 0
    mergeSort(J, 0, n - 1) # ジョイントを長さの昇順に整列
    J.append(0)            # 0 本使う場合
    totalj = 0
    for k in range(0, n):
        totalj += J[n - 1 - k]  # 長いものから選んでいく
        maxv = max((n - k) * (totalj + totalp), maxv)

    print(maxv)



